056-zk-kraft-migration.md

```diff
@@ -0,0 +1,254 @@
+
+In the `WaitBrokersInMigration` state, the operator expects the user to ask for running the actual metadata migration.
+In order to start the actual migration, the brokers need to be connected to the KRaft controllers and have the ZooKeeper migration enabled.
+The user annotates the `Kafka` custom resource running the current ZooKeeper based brokers with `strimzi.io/kraft-migration: migration`.
+This annotation drives the operator to update the brokers' configuration with the ZooKeeper migration flag enabled and with the new KRaft controllers quorum connection information as well.
+
+```shell
+# KRaft controller quorum configuration
+controller.quorum.voters=<voters>
+controller.listener.names=CONTROLLER
+
+# The controller.listener.name should also be added to listener.security.property.map
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,CONTROLLER:PLAINTEXT
+
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
+```
+
+Finally, the operator rolls the brokers.
+After the rolling, the internal Apache Kafka migration component starts the actual metadata migration and the operator has to wait for it to be completed to move to the next phase.
+Checking the migration progress can be done using some of the exposed migration metrics.
+The FSM moves to `Migration` state first, while migration is ongoing, and finally to the `DualWrite` state.
+In both states, the user can rollback the process by applying the `strimzi.io/kraft-migration: pre-migration` on the `Kafka` custom resource.
+In this case, the operator configures the brokers with no ZooKeeper migration flag, and no KRaft controllers quorum connection information.
+Finally, the operator rolls the brokers and wait for them to be ready again.
+The FSM moves back to the `WaitBrokersInMigration`, so actually at the beginning of the Phase 2.
+
+If the user doesn't rollback the migration, next step is to move brokers and controllers out of migration and using KRaft only in Phase 3/4.
+
+#### Phase 3/4: Migrating brokers and controllers to full KRaft
+
+In the `DualWrite` state, the operator expects the user desire to finalize the migration, having a full KRaft based cluster and ZooKeeper out of the picture.
+The user applies the `strimzi.io/kraft-migration: post-migration` annotation to the `Kafka` custom resource.
+The operator updates the brokers' configuration by removing the ZooKeeper migration flag and connection to the ZooKeeper ensemble.
+It also has to replace the `broker.id` with `node.id`.
+
+```shell
+# replace broker.id with node.id
+node.id=0
+
+# Remove the migration enabled flag
+# zookeeper.metadata.migration.enable=true
+
+# Remove ZooKeeper client configuration
+# zookeeper.connect=<connection-to-zookeeper>
+
+# Don't set the IBP, KRaft uses "metadata.version" feature flag
+# inter.broker.protocol.version=3.5
+```
+
+The operator rolls the brokers and waits for them to be ready again.
```

Comment:
You've said "rolls the broker" several times so far. Is it always just the brokers, and not the controllers that are restarted? It might be a good idea to be explicit in each case, just so its clear to the reader.

Reply:
In this case it's right. After being in "Dual Write" state, the first nodes to move out from the ZooKeeper usage are the brokers. If you follow the next step, then the controllers are rolled to move out from ZooKeeper usage and finalizing the migration. Everything should happen in the same reconcile loop. As you can see the section title is Phase 3/4. The first rejected proposal was envisaging two different steps: user getting out controllers first and still user acting for getting out brokers then. With this updated proposal is just one action/step for running both phases.